### ISSUE TITLE
restore.cmd crashes with exit code 255

### ISSUE THREAD

**Issue by Omcsesz** [dotnet/aspnetcore#33660]
### Describe the bug
Hello, Team!

When I'm running .\restore.cmd, the script crashes.

### To Reproduce
1. Download the repo
2. Download pre-requisites
3. Execute first step to build project (https://github.com/dotnet/aspnetcore/blob/main/docs/BuildFromSource.md  -> run .\restore.cmd)

### Exceptions (if any)
```
Building of C# project is enabled and has dependencies on NodeJS projects. Building of NodeJS projects is enabled since node is detected in C:\Program Files. Detected JDK in c:\Program Files\Java\jdk-15.0.1 (via JAVA_HOME)
Fatal error. System.Runtime.InteropServices.SEHException (0x80004005): External
component has thrown an exception.
Build failed with exit code 255. Check errors above.
```

### Further technical details
- **ASP.NET Core version** : 5.0.7
- **Include the output of `dotnet --info`**: [dotnetInfo.txt](https://github.com/dotnet/aspnetcore/files/6679781/dotnetInfo.txt)
- **The IDE (VS / VS Code/ VS4Mac) you're running on, and its version:** Microsoft Visual Studio Community 2019
Version 16.10.2

This is a gif containing the event of the error happening:
[restoreCmdCrashes](https://user-images.githubusercontent.com/1839739/122627104-d10ad600-d0ad-11eb-83a2-c664d8fbfda1.gif)

**Comment by dougbu** [comment]
@Omcsesz I wasn't able to reproduce this locally using our 'main' branch and a `.net6` SDK but that may not mean much. Could you please provide more information❔
- the branch or tag you are using
- whether the repo is clean e.g. have you run anything else in the repo and, if you did, did you execute `git clean -dix -e .dotnet/ -e .tools/` before trying `.\restore`
- the Windows version of your machine
- whether you used `eng/scripts/InstallVisualStudio.ps1 -edition Community` to ensure the installation has all necessary components

Worst case, the bug may be specific to your local environment. We might need `SET COREHOST_TRACE=1` output (perhaps redirected to a file w/ `SET COREHOST_TRACEFILE={path}`) to figure it out. But, let's not go there unless we need to.

----

@jeffhandley who owns System.Runtime.InteropServices❔ This is a native exception coming from somewhere in there.

@eerhardt @marek-safar any chance this problem is related to trimming features in dotnet/runtime❔

**Comment by Omcsesz** [comment]
* I'm using the main branch
* no, I have not run anything else
* I use Win7 [Version 6.1.7601]
* no, I haven't used eng/scripts/InstallVisualStudio.ps1 -edition Community

**Comment by eerhardt** [comment]
> @eerhardt @marek-safar any chance this problem is related to trimming features in dotnet/runtime❔

I would say that is pretty unlikely.

Getting a dump of the crashing process is your best bet here.

**Comment by Omcsesz** [comment]
I used the eng/scripts/InstallVisualStudio.ps1 -edition Community script since, but the issue persists.

**Comment by dougbu** [comment]
> Getting a dump of the crashing process is your best bet here.

Any specific `DOTNET_DbgMiniDumpType` setting @eerhardt❔

**Comment by eerhardt** [comment]
I would just leave it to the default to start. If more info is needed, we can get more later.

@Omcsesz - here are the docs on getting a dump: https://docs.microsoft.com/en-us/dotnet/core/diagnostics/dumps

**Comment by Omcsesz** [comment]
@eerhardt I created a dump file:

https://drive.google.com/file/d/1Z3GeE76o2rH4k2gd-ieDEgljzDv2NEfO/view?usp=sharing

**Comment by eerhardt** [comment]
The dump you sent was for powershell.exe running on .NET Framework. Is that the process that is crashing?

**Comment by Omcsesz** [comment]
That process opens a dotnet.exe file, for which I also created a dmp file:
https://drive.google.com/file/d/1UnNYtencuoIkpjmXrc6ARXK8k3jShufq/view?usp=sharing

**Comment by eerhardt** [comment]
Here's what I see:

![image](https://user-images.githubusercontent.com/8291187/122990406-861ce700-d369-11eb-914f-e9d3cb329e89.png)

It is running `msbuild aspnetcore\artifacts\toolset\restore.proj /t:__WriteToolsetLocation`, and the crash is coming from:

![image](https://user-images.githubusercontent.com/8291187/122990880-0a6f6a00-d36a-11eb-999e-43d07d9fe16e.png)

Which is weird because I don't see anything else in the call stack - no main, or what is calling `LocalizableStrings`.

![image](https://user-images.githubusercontent.com/8291187/122992642-f62c6c80-d36b-11eb-9354-62cfa28263fd.png)

But it does look like something to do with Workloads, since that is the class with the strings.

cc @dsplaisted - have you ever seen anything like this? I'm not sure why Workloads would even get in the picture here when running a basic msbuild command.

Adding some runtime folks who might be able to understand what is going on here: @AaronRobinsonMSFT @janvorli

**Comment by eerhardt** [comment]
@AaronRobinsonMSFT pointed me to https://github.com/dotnet/runtime/issues/54440, which looks very similar to this issue.

**Comment by eerhardt** [comment]
Since this is a Ready to run issue, you should be able to work around it by setting the environment variable `COMPlus_ReadyToRun=0`. And it will start working.

**Comment by Omcsesz** [comment]
@eerhardt after setting `COMPlus_ReadyToRun=0`, it passes through that point.
Thanks!
Will you fix this in some future release? Or update the build instructions?

**Comment by eerhardt** [comment]
Yes, the fix is in PR right now, and hopefully makes it into 6.0-preview6. But if it doesn't, it will definitely be fixed when 6.0 releases later this year.

Closing as a dupe of https://github.com/dotnet/runtime/issues/54440.

**Comment by dougbu** [comment]
> Yes, the fix is in PR right now

@AntonLapounov merged dotnet/runtime#54533 a couple of hours ago. Fix will flow into this repo and others automatically.

So far however, there's no in-flight backport to the 'release/6.0-preview6 branch in dotnet/runtime'. Can keep an eye on https://github.com/dotnet/runtime/pulls?q=is%3Apr+is%3Aopen+base%3Arelease%2F6.0-preview6 or chime in on dotnet/runtime#54533.

**Comment by mangod9** [comment]
since this is win7 specific issue we are not considering for backporting to preview6. Let us know if that isnt the case and we will rethink.
